src/terms/nonlocal.jl

```diff
@@ -75,17 +75,26 @@ end
             ψk = to_cpu(ψ[ik])
             occupationk = to_cpu(occupation[ik])
             form_factors = build_projector_form_factors(element.psp, G_plus_k_cart)
+
+            # Pre-allocation of large arrays
+            δHψk = similar(ψk)
+            P = similar(form_factors)
+            dPdR = similar(form_factors)
+            structure_factors = similar(form_factors, length(G_plus_k))
+            twoπp = similar(form_factors, length(G_plus_k))
+            
             for idx in group
                 r = model.positions[idx]
-                structure_factors = [cis2pi(-dot(p, r)) for p in G_plus_k]
```

Comment:
I bet we have other places, where we are missing this opportunity, too. Could you have a quick look @abussy ?